motionplan/nearestNeighbor.go

```diff
@@ -138,43 +124,29 @@ func (nm *neighborManager) nnWorker(ctx context.Context, planOpts *plannerOption
 	var best node
 	bestDist := math.Inf(1)
 
-	for {
-		select {
-		case <-ctx.Done():
-			nm.neighbors <- nil
-			return
-		default:
+	for candidate := range nm.nnKeys {
+		seg := &Segment{
+			StartConfiguration: nm.seedPos.Q(),
+			EndConfiguration:   candidate.Q(),
+		}
+		if pose := nm.seedPos.Pose(); pose != nil {
+			seg.StartPosition = pose
+		}
+		if pose := candidate.Pose(); pose != nil {
+			seg.EndPosition = pose
+		}
+		dist := planOpts.DistanceFunc(seg)
+		if dist < bestDist {
+			bestDist = dist
+			best = candidate
 		}
 
 		select {
-		case k := <-nm.nnKeys:
-			if k != nil {
-				nm.nnLock.RLock()
-				seg := &Segment{
-					StartConfiguration: nm.seedPos.Q(),
-					EndConfiguration:   k.Q(),
-				}
-				if pose := nm.seedPos.Pose(); pose != nil {
-					seg.StartPosition = pose
-				}
-				if pose := k.Pose(); pose != nil {
-					seg.EndPosition = pose
-				}
-				dist := planOpts.DistanceFunc(seg)
-				nm.nnLock.RUnlock()
-				if dist < bestDist {
-					bestDist = dist
-					best = k
-				}
-			}
+		case <-ctx.Done():
+			break
 		default:
-			nm.nnLock.RLock()
-			if nm.ready {
-				nm.nnLock.RUnlock()
-				nm.neighbors <- &neighbor{bestDist, best}
-				return
-			}
-			nm.nnLock.RUnlock()
 		}
 	}
+
+	nm.neighbors <- &neighbor{bestDist, best}
```

Comment:
The only functional change of this PR. In the old code -- if an `nnWorker` saw the context canceled, it would refuse to report its progress (returned a `nil`). And the nearest neighbor method gathering all the results from workers would simply ignore that `nil`. With this PR though I flipped that.

I was thinking about the question of what are the possible outcomes with the return values from `nnWorker`s falling into these 3 buckets:
1. All inputs are consumed, each nnWorker returns its version of the best answer.
2. The context is interrupted -- every nnWorker returns nil.
3. The context is interrupted -- some workers return their best candidate and some return nil.

I argue that (3) was only possible in an unhelpful way. I claim in that case, between all of the workers, every input was looked at. And the best value would have been returned by the workers that instead return `nil`. I'm not going to write down a proof, but I'm happy to go over that argument in more rigorous detail offline

I think the more interesting question is -- if the context is interrupted, should we return the "best so far" answer, or nothing to the top-level caller (and give the caller a way to distinguish between receiving the true answer vs our best approximation).

If we want to return nothing in that case, I'm happy to make a change.